### DISCUSSION TITLE
Freestyle libre protocol / creating a bangle js port

### DISCUSSION THREAD

**Discussion by Chriz76** [j-kaltes/Juggluco#158] [Q&A]
I guess you habe reverse engineered the freestyle libre protocol, the initial activating, pairing, encryption, etc. Do you hsve some documentation for it?

Otherwise can you give me a hint where to find the relevant classes in the source code? I guess I will have to activate a freestyle 3 with a modified original app and then copy the pairing secret to the watch to allow to communicate standalone.

That would give me the possibility to port Juggluco as a standalone app/watchface for bangle js. Which is much cheaper and has a much better battery life than wear os and is fully open for programming.

**Comment by j-kaltes** [comment]
Bangle js sounds like a kind of Fitbit on which you can only run Javascript. You will not easily connect it directly to Freestyle Libre sensors, but you can easily indirectly connect to the sensor via a smartphone running Juggluco. For example making use of the web server in Juggluco.
EDIT:
You could directly connect a sensor with an app running on [AsteriodOS](https://asteroidos.org/), but it has the same battery life as WearOS and runs on the same hardware.

**Reply by Chriz76** [reply]
Thank you @j-kaltes . Bangle js is fully open source including the bootloader. So I could even write c code but it won't be necesarry since all hw and the bluetooth stack can be accessed with javascript, too. 

The battery life is up 14 days. But will be less when having bluetooth turned on all the time. Still I need the protocol specifications of the freestyle. I might be able to reengineer it looking at the juggluco sources but hope that it is already somewhere documented.

**Comment by j-kaltes** [comment]
First try out how well the watch is able to receive the raw values from Sibionics sensors. That should be possible.

**Reply by Chriz76** [reply]
Thank you @j-kaltes. I could try that. The sensor is cheaper and no pairing needed? Is there an offical protocol description for the GS1 sensor or do you have a description in the wiki or a hint which juggluco files contain the relevant code?

**Comment by gui-dos** [comment]


**Reply by Chriz76** [reply]
Thank you so much. That's very impressive. It looks like I have to get hands on sensor to work on a real device. The bad thing is that this would mean to buy a sensor every 14 days...

**Comment by Chriz76** [comment]
I digged deeper into bangle js and juggluco and also have the opportunity to get some unused libre 2 sensors.

Bangle js can also be programmed via c and I already changed the firmware for another project. 

But the main problem seems to be the encryption of libre 2, 3 and even the gs1 sensors. As far as I could see juggluco uses for all three of them the binary libs from the original vendor apps. No chance to get them working on another platform.

@j-kaltes and @gui-dos is that assumption true or is maybe the gs1 unencrypted or libre 2,3 or the gs1 protocol fully reengineered?
